hnswlib/hnswalg.h

```diff
@@ -786,17 +768,39 @@ namespace hnswlib {
          * @param internalId
          */
         void markDeletedInternal(tableint internalId) {
-            unsigned char *ll_cur = ((unsigned char *)get_linklist0(internalId))+2;
-            *ll_cur |= DELETE_MARK;
+            if (!isMarkedDeleted(internalId))
+            {
+                unsigned char *ll_cur = ((unsigned char *)get_linklist0(internalId))+2;
+                *ll_cur |= DELETE_MARK;
+                num_deleted_ += 1;
+            }
+        }
+
+        /**
+         * Remove the deleted mark of the node, does NOT really change the current graph.
+         * @param label
+         */
+        void unmarkDelete(labeltype label)
+        {
+            auto search = label_lookup_.find(label);
+            if (search == label_lookup_.end()) {
+                throw std::runtime_error("Label not found");
+            }
+            tableint internalId = search->second;
+            unmarkDeletedInternal(internalId);
         }
 
         /**
          * Remove the deleted mark of the node.
          * @param internalId
          */
         void unmarkDeletedInternal(tableint internalId) {
-            unsigned char *ll_cur = ((unsigned char *)get_linklist0(internalId))+2;
-            *ll_cur &= ~DELETE_MARK;
+            if (isMarkedDeleted(internalId))
```

Comment:
I wonder should we throw an exception in case it is false and also check for the out-of-bounds?

Reply:
Added checks for internalID